tutorials/torchao_vllm_integration.md

```diff
@@ -0,0 +1,465 @@
+
+## Serialization and Model Sharing
+
+### SafeTensors Limitations
+
+**Important**: TorchAO quantized models currently cannot be serialized with safetensors due to tensor subclass limitations. SafeTensors only supports plain tensors, but TorchAO uses tensor subclasses with additional metadata that cannot be directly serialized.
+
+### Current Workaround
+
+When saving quantized models to HuggingFace Hub, you **must** use `safe_serialization=False`:
+
+```python
+from transformers import TorchAoConfig, AutoModelForCausalLM
+from torchao.quantization import Int4WeightOnlyConfig
+
+# Quantize model
+quantization_config = TorchAoConfig(Int4WeightOnlyConfig(group_size=128, use_hqq=True))
+model = AutoModelForCausalLM.from_pretrained(
+    "meta-llama/Llama-2-7b-hf",
+    quantization_config=quantization_config
+)
+
+# REQUIRED: Must use safe_serialization=False for tensor subclasses
+model.push_to_hub("your-username/llama-2-7b-int4", safe_serialization=False)
+```
+
+### The SafeTensors Challenge
+
+The fundamental issue is that tensor subclasses contain:
+
+1. **Multiple tensor components** (quantized data, scales, zero points)
+2. **Non-tensor metadata** (layouts, configurations, function references)
+3. **Complex reconstruction logic** that safetensors doesn't understand
+
+### Potential Solutions (Future Work)
+
+A proposed solution involves decomposing tensor subclasses for serialization:
+
+```python
+# Conceptual approach for future safetensors support
+def decompose_for_safetensors(tensor_subclass):
+    """Decompose tensor subclass into serializable components"""
+    # Extract plain tensors
+    plain_tensors = tensor_subclass.__tensor_flatten__()[0]
+
+    # Extract metadata
+    metadata = tensor_subclass.__tensor_flatten__()[1]
+
+    # Serialize function references as strings
+    serializable_metadata = {
+        k: serialize_if_function(v) for k, v in metadata.items()
+    }
+
+    return plain_tensors, serializable_metadata
+
+def reconstruct_from_safetensors(plain_tensors, metadata):
+    """Reconstruct tensor subclass from safetensors"""
+    # Deserialize function references
+    deserialized_metadata = {
+        k: deserialize_if_function(v) for k, v in metadata.items()
+    }
+
+    # Reconstruct tensor subclass
+    return TensorSubclass.__tensor_unflatten__(
+        plain_tensors, deserialized_metadata, None, None
+    )
+```
+
+However this is not quite ready yet.
+
+### Best Practices for Model Distribution
+
+1. **For Production**: Use `safe_serialization=False` when pushing quantized models
+2. **For Sharing**: Consider providing both quantized and unquantized versions
+3. **For Loading**: VLLM automatically handles tensor subclass reconstruction during model loading
+
+The TorchAO team is actively working on solutions to enable proper safetensors support for tensor subclasses.
+
+## Integration Architecture Diagrams
+
+### 1. High-Level Model Flow: Transformers → VLLM + TorchAO
```

Comment:
Add some short text here to talk about the high level? How are these diagrams related to each other?